File: saas/programming-model/compss-rt/bindings/python/apps/matmul/matmul.py
#
#  Copyright 2002-2014 Barcelona Supercomputing Center (www.bsc.es)
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


'''
@author: etejedor
'''

### HELPER FUNCTIONS ###

def initialize_variables():
    for i in range(MSIZE):
        A.append([])
        B.append([])
        C.append([])
        for j in range(MSIZE):
            A[i].append("A.%d.%d" % (i,j))
            B[i].append("B.%d.%d" % (i,j))
            C[i].append("C.%d.%d" % (i,j))

def fill_matrices():
    for c in ['A', 'B', 'C']:
        for i in range(MSIZE):
            for j in range(MSIZE):
                tmp = "%s.%d.%d" % (c,i,j)
                f = open(tmp, 'w')
                for _ in range(BSIZE):
                    for jj in range(BSIZE):
                        if c == 'C':
                            f.write('0.0')
                        else:
                            f.write('2.0')
                        if jj < BSIZE - 1:
                            f.write(' ')
                    f.write('\n')
                f.close()

def load_block(fi):
    b = []
    f = open(fi, 'r')
    for line in f:
        split_line = line.split(' ')
        r = []
        for num in split_line:
            r.append(float(num))
        b.append(r)
    f.close()
    return b

def store_block(b, fi, size):
    f = open(fi, 'w')
    for row in b:
        for j in range(size):
            num = row[j]
            f.write(str(num))
            if j < size - 1:
                f.write(' ')
        f.write('\n')  



### TASK SELECTION ###

from pycompss.api.task import task
from pycompss.api.parameter import *

@task(fa = FILE, fb = FILE, fc = FILE_INOUT)
def multiply(fa, fb, fc, size):
    a = load_block(fa)
    b = load_block(fb)
    c = load_block(fc)
    for i in range(size):
        for j in range(size):
            for k in range(size):
                c[i][j] += a[i][k] * b[k][j];
    store_block(c, fc, size)



### MAIN PROGRAM ###

if __name__ == "__main__":
    
    import sys
    
    args = sys.argv[1:]
    
    MSIZE = int(args[0])
    BSIZE = int(args[1])
    
    A = []
    B = []
    C = []
    
    initialize_variables()
    fill_matrices()
    
    for i in range(MSIZE):
        for j in range(MSIZE):
            for k in range(MSIZE):
                multiply(A[i][k], B[k][j], C[i][j], BSIZE)
    

File: saas/programming-model/compss/compss-rt/bindings/c/apps/matmul/matmul.cc
/*
 *  Copyright 2002-2014 Barcelona Supercomputing Center (www.bsc.es)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */



#include <time.h>
#include <stdio.h>
#include <errno.h>
#include "matmul.h"
#include "GS_compss.h"

#include "parameters.h"


int main(int argc, char **argv)
{
	long int t = time(NULL);
	char *f1;
	char *f2;
	char *f3;
	char *file;
	FILE *fp;
	char c;
	int i, j, k, ii, jj;

	f1 = (char *)malloc(sizeof(char)*15);
	f2 = (char *)malloc(sizeof(char)*15);
	f3 = (char *)malloc(sizeof(char)*15);
	file = (char *)malloc(sizeof(char)*15);

	GS_On();

	for (c = 'A'; c < 'D'; c++) {
		for (i = 0; i < MSIZE; i++) {
			for (j = 0; j < MSIZE; j++) {
				sprintf(file, "%c.%d.%d", c, i, j);
				fp = fopen(file, "w");
				for (ii = 0; ii < BSIZE; ii++) {
					for (jj = 0; jj < BSIZE; jj++) {
						if (c == 'C') {
							fprintf(fp, "%lf ", 0.0);
						} else {
							fprintf(fp, "%lf ", 2.0);
						}
					}
					fprintf(fp, " \n");
				}
				fclose(fp);
			}
		}
	}

	for (i = 0; i < MSIZE; i++) {
		for (j = 0; j < MSIZE; j++) {
			for (k = 0; k < MSIZE; k++) {
				sprintf(f1, "A.%d.%d", i, k);
				sprintf(f2, "B.%d.%d", k, j);
				sprintf(f3, "C.%d.%d", i, j);
				matmul(f1, f2, &f3);
			}
		}
	}

	GS_Off();

	free(f1);
	free(f2);
	free(f3);
	free(file);

	printf("Total time:\n");
	t = time(NULL) - t;
	printf("%li Hours, %li Minutes, %li Seconds\n", t / 3600, (t % 3600) / 60,
			 (t % 3600) % 60);

	return 0;
}
